STEEL-95 resource_cwf_process improvements

crunis opened the pull request:
WIP - Still needs tests and documentation

raphael commented on rightscale/rsc/client.go:
```diff
@@ -722,3 +729,29 @@ func (inFields Fields) onlyPopulated() Fields {
+func analyzeSource(source string) (expectsOutputs bool, err error) {
+	const validDefinition = "^[[:blank:]]*define[[:blank:]]*[\\w_\\.]+[[:blank:]]*\\([@$\\w _,]*\\)[[:blank:]]+(return.*)?do"
+
+	r, _ := regexp.Compile(validDefinition)
```
Use `MustCompile` instead of ignoring the error. In general ignoring errors is a big no no. I didn't comment on line 662 because we know the value was previously a JSON value. But even there you could consider handling the error.

raphael replied:
Probably doesn't change much in practice since it would be invalid syntax but to be more precise the regexp could include a blank character class after the `return`.

raphael commented on rightscale/rsc/client.go:
```diff
@@ -722,3 +729,29 @@ func (inFields Fields) onlyPopulated() Fields {
+		expectsOutputs = true
+	}
+
+	return expectsOutputs, nil
```
The last 8 lines could just be `return matched[1] != "", nil`

raphael commented:
Build should be fixed before merging, would be good to understand why it failed (haven't looked).

crunis commented:
@raphael do you mind doing final Peer Review ? Latest changes:

1. Added documentation (https://github.com/rightscale/terraform-provider-rightscale/blob/19d07411cafd8d0be4dcd4fed36d457312810389/website/docs/r/cwf_process.markdown)
2. Added timeout while expecting Outputs (currently times out 10s AFTER the CFW process has finished). Covers the scenario where the user specifies an output in the return clause but forgets to fill in that value (without this timeout, the rsc client would wait forever for the outputs, TF would probably timeout after 10 minutes). See https://github.com/rightscale/terraform-provider-rightscale/pull/23/commits/25244ec65a725397a7f33ddc46bf0ab1ecf23542
3. Remember that Ive changed the behaviour with CWF errors, instead of filling up the status and error outputs, we return a TF error that the user immediately receives, so the user doesn't miss the error or needs to keep checking for errors on his TF CWF declarations. See https://github.com/rightscale/terraform-provider-rightscale/pull/23/commits/ad06b3b48bda13ec19af7d461387886026095fe1

raphael commented on rightscale/rsc/client.go:
```diff
@@ -655,7 +662,13 @@ func processOutputs(res interface{}) map[string]interface{} {
 		om := out.(map[string]interface{})
-		outputs[om["name"].(string)] = om["value"].(map[string]interface{})["value"]
+		value := om["value"].(map[string]interface{})["value"]
+		stringValue, ok := value.(string)
```
Go pro-tip: do not use mix case local variable names. Local variable names should be "short" (ideally one or two characters).

raphael commented on rightscale/rsc/client.go:
```diff
@@ -722,3 +735,23 @@ func (inFields Fields) onlyPopulated() Fields {
+func analyzeSource(source string) (expectsOutputs bool, err error) {
+	const validDefinition = "^[[:blank:]\n]*define[[:blank:]]*[\\w_\\.]+[[:blank:]]*\\([@$\\w _,]*\\)[[:blank:]]+(return.*[[:blank:]]+)?do"
+
+	r := regexp.MustCompile(validDefinition)
```
The above could be a global variable so the regex doesn't get recompiled for each invocation (probably doesn't change much in reality).

raphael commented on rightscale/rsc/client_test.go:
```diff
@@ -290,3 +290,47 @@ func TestOnlyPopulated(t *testing.T) {
+	}
+
+	for _, tt := range sourcetests {
+		t.Run(tt.source, func(t *testing.T) {
```
You probably want to give each test a name rather than use the source as name (add a initial field in the struct `name string`).

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+
+Use this resource to create or destroy RightScale [CloudWorkFlow processes](http://docs.rightscale.com/ss/reference/rcl/).
+
+Creating the CWF process will run it synchronously and return the output values (if any).
```
Use the present tense: `Creating a CWF process runs it synchronously and returns ...`

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+
+Creating the CWF process will run it synchronously and return the output values (if any).
+
+Destroying a CWF process will remove it from the CWF Console table of executed processes. It WON'T undo any action that it could have executed nor would stop any running process.
```
The fact that processes are listed in the CWF Console is ancillary really. The description should just specify that the process gets deleted. Also deleting a process does cause it to stop (in error).

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+
+## Example Usage
+
+This example CWF process will look for all servers whose name starts by "db-slave-" and will execute the specified RightScript on them,
```
will look -> looks
whose name starts by -> whose names start with
will execute -> executes

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+## Example Usage
+
+This example CWF process will look for all servers whose name starts by "db-slave-" and will execute the specified RightScript on them,
+returning the number of servers that have been reached.
```
reached -> affected

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+  parameters = [
+     { "kind" = "string"
+       "value" = "db-slave-" },
+     { "kind" = "string"
```
missing `,`

crunis replied:
It's correct, TF maps don't separate key-values with comas, the above example has been tested and it's like this in [official examples](https://www.terraform.io/docs/configuration/variables.html#example)

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+define main($server_prefix, $rightscript_href) return $servers_reached do
+  @servers = rs_cm.servers.get(filter: ["name==" + $server_prefix])
+  @servers.current_instance().run_executable(right_script_href: $rightscript_href)
+  $servers_reached = size(@servers)
```
don't think this works: not all servers may have current instances.

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,84 @@
+
+The following arguments are supported:
+
+* `source` - (Required) Source code to be executed, written in [RCL (RightScale CloudWorkFlow Language)](http://docs.rightscale.com/ss/reference/rcl/v2/index.html). The function should always be called main. Example:
```
Can there be more than one functions? specify it if so.

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -0,0 +1,88 @@
+
+Use this resource to create or destroy RightScale [CloudWorkFlow processes](http://docs.rightscale.com/ss/reference/rcl/).
+
+Creating the CWF process runs it synchronously and returns the output values (if any). If the CWF process fails, the Terraform script will fail too.
```
The Terraform scripts fails too.

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -10,16 +10,16 @@ description: |-
+Creating the CWF process runs it synchronously and returns the output values (if any). If the CWF process fails, the Terraform script will fail too.
 
-Destroying a CWF process will remove it from the CWF Console table of executed processes. It WON'T undo any action that it could have executed nor would stop any running process.
+Destroying the resource deletes the corresponding CWF process. If the process is still running, destroying it will make it end with an error.
```
Destroying a running process causes it to end in error.

raphael commented on website/docs/r/cwf_process.markdown:
```diff
@@ -10,16 +10,16 @@ description: |-
 
-This example CWF process will look for all servers whose name starts by "db-slave-" and will execute the specified RightScript on them,
-returning the number of servers that have been reached.
+This example CWF process looks for all servers whose names start by "db-slave-" and executes the specified RightScript on them,
```
start with

crunis commented:
Tiny nitpicks addressed! do you approve ?

raphael commented:
LGTM
